src/Database/LSMTree/Internal.hs

```diff
@@ -403,39 +417,63 @@ mkLevelsCache lvls = LevelsCache_ {
 -- Implementation of public API
 --
 
+{-# SPECIALISE withTable :: Session IO h -> TableConfig -> (TableHandle IO h -> IO a) -> IO a #-}
+-- | See 'Database.LSMTree.Normal.withTable'.
+withTable ::
+     m ~ IO -- TODO: replace by @io-classes@ constraints for IO simulation.
+  => Session m h
+  -> TableConfig
+  -> (TableHandle m h -> m a)
+  -> m a
+withTable sesh conf = bracket (new sesh conf) close
+
 {-# SPECIALISE new :: Session IO h -> TableConfig -> IO (TableHandle IO h) #-}
 -- | See 'Database.LSMTree.Normal.new'.
 new ::
      m ~ IO -- TODO: replace by @io-classes@ constraints for IO simulation.
   => Session m h
   -> TableConfig
   -> m (TableHandle m h)
-new sesh conf = do
+new sesh conf = newWithLevels sesh conf (\_ -> pure V.empty)
+
+{-# SPECIALISE new :: Session IO h -> TableConfig -> IO (TableHandle IO h) #-}
+newWithLevels ::
+     m ~ IO -- TODO: replace by @io-classes@ constraints for IO simulation.
+  => Session m h
+  -> TableConfig
+  -> (forall a. HasFS m h -> Cont (m a) (V.Vector (Level (Handle h))))
+     -- ^ Resources have to be cleaned up correctly if an (asynchronous)
+     -- exception occurs. We achieve this by requiring that the resource acquire
+     -- action is a 'Cont'inuation. See 'openRun' for more information.
+  -> m (TableHandle m h)
+newWithLevels sesh conf acqLevels = do
```

Comment:
I think we can re-use the code without having to parametrise by a function. Currently, the structure is roughly:

```haskell
new = newWithLevels (const empty)
open = newWithLevels openLevels
newWithLevels acq = do
  withOpenSession $ do
    runCont acq $ \levels ->
      _  -- create table

```

But this seems a little simpler:

```haskell
new =
  withCont $ do
    createTables empty
open = do
  withOpenSession $
    runCont acq $ \levels ->
      createTable levels
createTable levels = _  -- create table
```

Reply:
That indeed simplified the code a bit. Thanks!